### PR TITLE
[WIP] Add OpenAI compatibility to the bot

### PR DESCRIPTION
**Pull request by Soapy7261**
This checks for if the token starts with "gsk_" (groq) or "sk-" (OpenAI) and switches models and API URLs accordingly automatically, all you have to do is specify a different token!
Marked as a draft because I haven't tested this yet even though theoretically nothing *should* break

### CODE REVIEW

**Comment by Soapy7261** [comment]
also disables all AI features if an invalid token has been passed in commit 3

**Comment by Soapy7261** [comment]
~~I couldn't find a place where `generate_single` is used, so i commented it out~~

File: utils/api_helpers.py [hunk 1]
```diff
@@ -35,27 +61,38 @@ async def analyse_image(image_url: str) -> str:
...
+async def generate_single(prompt: str) -> str: # For future me, this is never used as of 2024-12-21 YYYY-MM-DD
+    client, language_model, vision_model = get_client()
+    try:
+        chat_completion = await client.chat.completions.create(messages=[{"role": "user", "content": prompt}],
+                                                            model=language_model, max_tokens=400)
+    except Exception as err:
+        log.error(
+            f"An error occurred while generating single: {err}",
+            exc_info=(type(err), err, err.__traceback__))
+        return "An error occurred. Please try again later, this has been logged."
```
Comment (ToothyDev) [inline utils/api_helpers.py]: Remove this error handling too, it's handled in the global error handler

File: example.config.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+token = "YOUR_DISCORD_BOT_TOKEN_HERE" # Obviously you need to specify this.
+groq_api_key = "YOUR_GROQ_OR_OPENAI_API_KEY" # Will disable the AI functions if not provided
```
Comment (ToothyDev) [inline example.config.py]: Rename this to ai_api_key or something similar, since it's not groq only anymore

File: cogs/utility.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import json
 import random
+import config # Help?
```
Comment (ToothyDev) [inline cogs/utility.py]: Unnecessary import

File: utils/api_helpers.py [hunk 4]
```diff
@@ -1,22 +1,48 @@
 import aiohttp
-from groq import AsyncGroq
+from groq import AsyncGroq as ai
```
Comment (ToothyDev) [inline utils/api_helpers.py]: Import it as AI, for consistency

File: utils/api_helpers.py [hunk 5]
```diff
@@ -1,22 +1,48 @@
...
 
 import config
+import logger
 from utils.data import Fursona
 
-LANGUAGE_MODEL = "llama-3.3-70b-versatile"
-VISION_MODEL = "llama-3.2-11b-vision-preview"
+log = logger.get_logger(__name__)
```
Comment (ToothyDev) [inline utils/api_helpers.py]: Remove the logger because I don't think it's needed anymore with my proposed changes

File: cogs/utility.py [hunk 6]
```diff
@@ -23,6 +25,8 @@ def __init__(self, bot: discord.Bot):
             required=False)
     async def sonagen(self, ctx: discord.ApplicationContext, species: str, sex: str, sonatype: str):
         """ Generate a random sona """
+        if not self.ai_enabled:
+            return await ctx.respond("AI functions are disabled due to missing (or invalid) API key, please contact the bot owner to fix this.")
```
Comment (ToothyDev) [inline cogs/utility.py]: Make the response ephemeral, just say "AI functions are currently disabled" without the last part and log it instead (look at how I log stuff elsewhere in the bot)

File: cogs/socials.py [hunk 7]
```diff
@@ -234,11 +235,13 @@ async def gay(self, ctx, user=None, border=False, server_avatar=True):
     @option("text", str, description="What do you want to tell Paw?")
     async def gpt(self, ctx: discord.ApplicationContext, text: str):
         """ Talk to Paw! """
+        if not self.ai_enabled:
+            return await ctx.respond("AI functions are disabled due to missing (or invalid) API key, please contact the bot owner to fix this.")
```
Comment (ToothyDev) [inline cogs/socials.py]: Make the response ephemeral, just say "AI functions are currently disabled" without the last part and log it instead (look at how I log stuff elsewhere in the bot)

**Review by ToothyDev — CHANGES_REQUESTED** [review]

File: utils/api_helpers.py (hunk 1 above)
Comment (Soapy7261) [reply utils/api_helpers.py]: My bad, I thought i removed it earlier.

**Comment by Soapy7261** [comment]
Your new config has text_model and vision_model hard coded to use groq, while someone could update them to use OpenAI models, I think it'd be best to either

A) Have some way to either check if it's using an OpenAI token but a groq model (eg, llama3.3) or a groq token but an OpenAI model (eg, gpt-4o), and disable all AI functions if so, could be done although who knows what OpenAI is going to name their models in the future.

B) Have `groq_text_model` and `openai_text_model` in the config

C) Have the text_model have a seperator for groq and OpenAI, like `text_model="llama-3.3-70b-versatile|gpt-4o-mini-2024-07-18"` which might be more confusing but gets rid of the need for 2 config entries

**Comment by Soapy7261** [comment]
Also i still need to run the bot to make sure it works at all xd

**Comment by ToothyDev** [comment]
> Your new config has text_model and vision_model hard coded to use groq, while someone could update them to use OpenAI models, I think it'd be best to either
> 
> A) Have some way to either check if it's using an OpenAI token but a groq model (eg, llama3.3) or a groq token but an OpenAI model (eg, gpt-4o), and disable all AI functions if so, could be done although who knows what OpenAI is going to name their models in the future.
> 
> B) Have `groq_text_model` and `openai_text_model` in the config
> 
> C) Have the text_model have a seperator for groq and OpenAI, like `text_model="llama-3.3-70b-versatile|gpt-4o-mini-2024-07-18"` which might be more confusing but gets rid of the need for 2 config entries

The example config is supposed to just be an example, and I set it to the llama models because that's also what I run the bot with. I assume people who use either OpenAI or Groq are smart enough to also change the models accordingly, if they already manage to get the bot running in the first place. I'll leave it like this because it provides a working default provided you enter a groq key, which, again, is the assumed default as of now

File: example.config.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
-token = "YOUR_DISCORD_BOT_TOKEN_HERE" # Obviously you need to specify this.
-groq_api_key = "YOUR_GROQ_OR_OPENAI_API_KEY" # Will disable the AI functions if not provided
+token = "BOT_TOKEN"
+ai_api_key = "GROQ_API_KEY"
```
Comment (ToothyDev) [inline example.config.py:2]: Name this something that makes it apparent that both a groq and openai key can be entered

File: example.config.py [hunk 9]
```diff
@@ -1,5 +1,5 @@
 token = "BOT_TOKEN"
-groq_api_key = "GROQ_API_KEY"
+ai_api_key = "GROQ_API_KEY"
```
Comment (ToothyDev) [inline example.config.py]: Name this something that makes it apparent that both a groq and openai key can be entered

**Review by ToothyDev — CHANGES_REQUESTED** [review]

**Comment by Soapy7261** [comment]
I believe that should be what you want

**Review by ToothyDev — APPROVED** [review]

**Comment by Soapy7261** [comment]
I'll add a few changes before i mark this as ready, won't take long (hopefully)

**Comment by Soapy7261** [comment]
For future me, this is closed because its pretty much already implemented, I'll make another PR for the other things i was gonna add

**Pull request closed without merging**